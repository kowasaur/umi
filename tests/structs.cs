struct Rectangle {
    public int length;
    public int width;
}
