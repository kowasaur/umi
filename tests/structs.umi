class Thing : Struct {
    mut Int number
    Int other

    static String sep

    Thing(Int n, Int o) {
        number = n
        other = o
    }

    static Thing() { sep = ", " }

    Void print() {
        print(number.toString() + sep + other.toString())
    }

    static Void staticMethod() { print("Static methods work as well") }
}

Void add1AndPrint(Thing thing) {
    thing.number = thing.number + 1
    thing.print()
}

Void main() {
    Thing thing = Thing(6, 5)
    thing.number = 10

    add1AndPrint(thing) # This does not modify thing since it is passed by value
    thing.print()

    print(Thing.sep)
    Thing.staticMethod()

    # You can assign a variable to itself in its definition to make it the default values
    Thing thing2 = thing2
    thing2.print()
}
