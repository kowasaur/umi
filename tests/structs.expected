11, 5
10, 5
, 
Static methods work as well
0, 0
